### PR TITLE
Adding support for take_along_axis and put_along_axis

### PR DESCRIPTION
**Pull request by ipdemes**

### CODE REVIEW

File: tests/integration/test_put_along_axis.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+    np_indices = mk_seq_array(np, shape_idx) % 5
+    num_indices = mk_seq_array(num, shape_idx) % 5
+    for axis in range(ndim):
+        res_np = np.put_along_axis(np_arr, np_indices, 8, axis=axis)
+        res_num = num.put_along_axis(num_arr, num_indices, 8, axis=axis)
+        assert np.array_equal(res_num, res_np)
+    np_indices = mk_seq_array(np, (3,))
+    num_indices = mk_seq_array(num, (3,))
+    res_np = np.put_along_axis(np_arr, np_indices, 11, None)
+    res_num = num.put_along_axis(num_arr, num_indices, 11, None)
```
Comment (bryevdv) [inline tests/integration/test_put_along_axis.py:65]: What do these statements after the loop do? I don't see an assertion that uses these values, so they seem to be superfluous. BTW the  `for axis in range(ndim)` loop could also be moved out to `pytest.parametrize`. I haven't done it previously since I think we would need to add `pytest-lazy-fixture` as a test dep, but let me know if you'd like to look into it.

File: tests/integration/test_take_along_axis.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+    np_indices = mk_seq_array(np, shape_idx) % 5
+    num_indices = mk_seq_array(num, shape_idx) % 5
+    for axis in range(ndim):
+        res_np = np.take_along_axis(np_arr, np_indices, axis=axis)
+        res_num = num.take_along_axis(num_arr, num_indices, axis=axis)
+        assert np.array_equal(res_num, res_np)
+    np_indices = mk_seq_array(np, (3,))
+    num_indices = mk_seq_array(num, (3,))
+    res_np = np.take_along_axis(np_arr, np_indices, None)
+    res_num = num.take_along_axis(num_arr, num_indices, None)
```
Comment (bryevdv) [inline tests/integration/test_take_along_axis.py:65]: same question as above, these last compute values just seem to get thrown away

File: tests/integration/test_put_along_axis.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+
+@pytest.mark.parametrize("ndim", range(1, LEGATE_MAX_DIM + 1))
+def test_ndim(ndim):
+    shape = (5,) * ndim
+    np_arr = mk_seq_array(np, shape)
+    num_arr = mk_seq_array(num, shape)
+    shape_idx = (1,) * ndim
+    np_indices = mk_seq_array(np, shape_idx) % 5
+    num_indices = mk_seq_array(num, shape_idx) % 5
```
Comment (bryevdv) [inline tests/integration/test_put_along_axis.py:57]: Is the value 5 here coordinated with the `(5,) * ndim` above (and also similar code in the test below)? If so, suggest pulling the value out into a module level "const" e.g. `N = 5` (or whatever var name is appropriate) that is referred to in all these different places

File: cunumeric/module.py [hunk 4]
```diff
@@ -2498,6 +2499,141 @@ def take(
...
+        raise ValueError(
+            "`indices` and `a` must have the same number of dimensions"
+        )
+    return a[_fill_fancy_index_for_along_axis_routines(a.shape, axis, indices)]
+
+
+add_boilerplate("a", "indices", "values")
+
+
+def put_along_axis(
```
Comment (bryevdv) [inline cunumeric/module.py:2579]: function decorators usually don't have vertical whitespace between

File: cunumeric/module.py [hunk 5]
```diff
@@ -2498,6 +2499,140 @@ def take(
     return a.take(indices=indices, axis=axis, out=out, mode=mode)
 
 
+def _fill_fancy_index_for_along_axis_routines(
+    a_shape: tuple, axis: int, indices: ndarray
```
Comment (bryevdv) [inline cunumeric/module.py:2503]: is this specifically `tuple[int, ...]`? If so NdShape (in `.types`) can be used for that

File: cunumeric/module.py [hunk 6]
```diff
@@ -2498,6 +2499,140 @@ def take(
...
+) -> tuple:
```
Comment (bryevdv) [inline cunumeric/module.py:2504]: Also needs full type spec, `tuple[<something>]` (or `NdShape`  if that is applicable)

File: cunumeric/module.py [hunk 7]
```diff
@@ -2498,6 +2499,140 @@ def take(
...
+    Multiple GPUs, Multiple CPUs
+    """
+    if not np.issubdtype(indices.dtype, np.integer):
+        raise IndexError("`indices` must be an integer array")
+
+    if axis is None:
+        a = a.ravel()
+        axis = 0
+        if indices.ndim != 1:
+            raise ValueError("indices must be 1D if axis=None")
```
Comment (bryevdv) [inline cunumeric/module.py:2564]: This check doesn't depend on `a` or `axis` so I'd advise making it the very first thing at the top of the conditional block.

File: cunumeric/module.py [hunk 8]
```diff
@@ -2498,6 +2499,140 @@ def take(
...
+
+    """
+    if not np.issubdtype(indices.dtype, np.integer):
+        raise IndexError("`indices` must be an integer array")
+
+    if axis is None:
+        a = a.ravel()
+        axis = 0
+        if indices.ndim != 1:
+            raise ValueError("indices must be 1D if axis=None")
```
Comment (bryevdv) [inline cunumeric/module.py:2623]: same comment, bail out as early as possible (before any unnecessary work is done) when feasible

File: cunumeric/module.py [hunk 9]
```diff
@@ -2498,6 +2499,140 @@ def take(
...
+    else:
+        axis = normalize_axis_index(axis, a.ndim)
+
+    if a.ndim != indices.ndim:
+        raise ValueError(
+            "`indices` and `a` must have the same number of dimensions"
+        )
+    a[
+        _fill_fancy_index_for_along_axis_routines(a.shape, axis, indices)
+    ] = values
```
Comment (bryevdv) [inline cunumeric/module.py:2633]: This is a case where black's formatting made things hard to parse (at a glance the call to `_fill_fancy_index_for_along_axis_routines` looked like it was under the `raise`) Suggest saving off the computed value:
```suggestion
    ind = _fill_fancy_index_for_along_axis_routines(a.shape, axis, indices)
    a[ind] = values
```
This also makes it easier in the future if anyone wants to put a print debug or debugger statement in to examine the value of the computed inds.

File: tests/integration/test_put_along_axis.py (hunk 1 above)
Comment (ipdemes) [reply tests/integration/test_put_along_axis.py:68]: you are right, there is an assertion missing

File: cunumeric/module.py (hunk 6 above)
Comment (ipdemes) [reply cunumeric/module.py:2504]: thanks! I am still learning how to correctly do type annotation. It should be tuple of ndarrays

File: cunumeric/module.py (hunk 9 above)
Comment (ipdemes) [reply cunumeric/module.py:2633]: makes sense!

File: cunumeric/module.py (hunk 4 above)
Comment (bryevdv) [reply cunumeric/module.py:2579]: Oh, I think an `@` is missing in front of `add_boilerplate`?

Comment (ipdemes) [reply cunumeric/module.py:2579]: Yes! thank you!

File: cunumeric/module.py [hunk 10]
```diff
@@ -2498,6 +2499,139 @@ def take(
...
+    See Also
+    --------
+    numpy.take_along_axis
+
+    Availability
+    --------
+    Multiple GPUs, Multiple CPUs
+    """
+    if not np.issubdtype(indices.dtype, np.integer):
+        raise IndexError("`indices` must be an integer array")
```
Comment (manopapad) [inline cunumeric/module.py:2559]: I think this should be `TypeError`.

File: cunumeric/module.py [hunk 11]
```diff
@@ -2498,6 +2499,139 @@ def take(
...
+    Multiple GPUs, Multiple CPUs
+
+    """
+    if not np.issubdtype(indices.dtype, np.integer):
+        raise IndexError("`indices` must be an integer array")
+
+    if axis is None:
+        if indices.ndim != 1:
+            raise ValueError("indices must be 1D if axis=None")
+        a = a.ravel()
```
Comment (manopapad) [inline cunumeric/module.py:2622]: This can return a copy, meaning the user-supplied `a` will end up not being modified. We should check here that `a.ravel()` indeed created a view. I created a PR adding such a method: https://github.com/nv-legate/cunumeric/pull/474.

Alternatively, we can require that if `axis` is missing then `a` must be 1d (and update the documentation accordingly). Any preference @magnatelee ?

For what it's worth, it looks like the current implementation will create a copy even if reshape just drops a singleton dimension (if run with `-cunumeric:test`, so we go through `DeferredArray`):

```
a = cn.ones((12,1))
b = a.ravel()
print(a.same_root(b))  # prints `False`
```

File: cunumeric/module.py [hunk 12]
```diff
@@ -2498,6 +2499,139 @@ def take(
...
+    --------
+    numpy.put_along_axis
+
+    Availability
+    --------
+    Multiple GPUs, Multiple CPUs
+
+    """
+    if not np.issubdtype(indices.dtype, np.integer):
+        raise IndexError("`indices` must be an integer array")
```
Comment (manopapad) [inline cunumeric/module.py:2617]: This should be `TypeError`.

File: cunumeric/module.py [hunk 13]
```diff
@@ -2498,6 +2499,139 @@ def take(
...
+        Indices to change along each 1d slice of `arr`. This must match the
+        dimension of arr, but dimensions in Ni and Nj may be 1 to broadcast
+        against `arr`.
+    values : array_like (Ni..., J, Nk...)
+        values to insert at those indices. Its shape and dimension are
+        broadcast to match that of `indices`.
+    axis : int
+        The axis to take 1d slices along. If axis is None, the destination
+        array is treated as if a flattened 1d view had been created of it.
+
```
Comment (manopapad) [inline cunumeric/module.py:2606]: Let's add a note here, that having duplicate entries in `indices` will result in undefined behavior. This is also technically the case with vanilla NumPy, see https://stackoverflow.com/questions/15973827/handling-of-duplicate-indices-in-numpy-assignments, but the NumPy documentation doesn't advertise one way or the other.

File: tests/integration/test_take_along_axis.py [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+
+    indices = mk_seq_array(np, (256, 256, 10)) % 100
+    indices_num = num.array(indices)
+
+    res = np.take_along_axis(x, indices, -1)
+    res_num = num.take_along_axis(x_num, indices_num, -1)
+    assert np.array_equal(res_num, res)
+
+
+def test_None_axis():
```
Comment (manopapad) [inline tests/integration/test_take_along_axis.py:37]: Isn't this case already checked in `test_ndim`?

File: tests/integration/test_put_along_axis.py [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+@pytest.mark.parametrize("ndim", range(1, LEGATE_MAX_DIM + 1))
+def test_ndim(ndim):
+    shape = (N,) * ndim
+    np_arr = mk_seq_array(np, shape)
+    num_arr = mk_seq_array(num, shape)
+    shape_idx = (1,) * ndim
+    np_indices = mk_seq_array(np, shape_idx) % N
+    num_indices = mk_seq_array(num, shape_idx) % N
+    for axis in range(ndim):
+        res_np = np.put_along_axis(np_arr, np_indices, 8, axis=axis)
```
Comment (manopapad) [inline tests/integration/test_put_along_axis.py:62]: `put_along_axis` doesn't return anything. You need to create a new `np_arr/num_arr` each time and compare them after `put_along_axis` has updated them.

File: tests/integration/test_put_along_axis.py [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+        res_num = num.put_along_axis(num_arr, num_indices, 8, axis=axis)
+        assert np.array_equal(res_num, res_np)
+    np_indices = mk_seq_array(np, (3,))
+    num_indices = mk_seq_array(num, (3,))
+    res_np = np.put_along_axis(np_arr, np_indices, 11, None)
```
Comment (manopapad) [inline tests/integration/test_put_along_axis.py:67]: Same comment as above.

File: tests/integration/test_put_along_axis.py [hunk 17]
```diff
@@ -0,0 +1,75 @@
...
+
+    indices = mk_seq_array(np, (256, 256, 10)) % 100
+    indices_num = num.array(indices)
+
+    res = np.put_along_axis(x, indices, -10, -1)
+    res_num = num.put_along_axis(x_num, indices_num, -10, -1)
+    assert np.array_equal(res_num, res)
+
+
+def test_None_axis():
```
Comment (manopapad) [inline tests/integration/test_put_along_axis.py:37]: I think this case is covered in `test_ndim` below.

File: tests/integration/test_put_along_axis.py [hunk 18]
```diff
@@ -0,0 +1,75 @@
...
+
+def test_3d():
+
+    x = mk_seq_array(np, (256, 256, 100))
+    x_num = mk_seq_array(num, (256, 256, 100))
+
+    indices = mk_seq_array(np, (256, 256, 10)) % 100
+    indices_num = num.array(indices)
+
+    res = np.put_along_axis(x, indices, -10, -1)
```
Comment (manopapad) [inline tests/integration/test_put_along_axis.py:32]: Here also compare `x` after the updates, not the return value of `put_along_axis`.

File: cunumeric/module.py (hunk 11 above)
Comment (ipdemes) [reply cunumeric/module.py:2947]: Numpy documentation doesn't say it has to return the view, but I believe it does actually return the view because it uses `a.flat`
Cupy, however, calls a.ravel().
I think we can still support the case when axis=None and a.ndim>0 returning a copy  as far as we document this. Supporting some case partially can be better than not supporting it at all.

File: tests/integration/test_take_along_axis.py (hunk 14 above)
Comment (ipdemes) [reply tests/integration/test_take_along_axis.py:37]: yes, it is. As well as test_3d is tested in test_ndim

File: cunumeric/module.py (hunk 11 above)
Comment (manopapad) [reply cunumeric/module.py]: > Numpy documentation doesn't say it has to return the view

On the NumPy documentation I see: "If axis is None, the destination array is treated as if a flattened 1d *view* had been created of it." So it looks like NumPy guarantees a view even if `axis` is `None`.

> I think we can still support the case when axis=None and a.ndim>1 returning a copy

The function has no return value, its only effect is updating the passed array.

Comment (ipdemes) [reply cunumeric/module.py]: Yes, you are right, I didn't check the lines and though this comment is for `take_along_axis`. I can disable this use case until we get  `a.flat` implemented

Comment (manopapad) [reply cunumeric/module.py]: Yes, let's error out for now when `axis is None` and `a.ndim > 1`. But please leave a TODO note, to revisit this once #474 or something equivalent is merged.

File: cunumeric/module.py [hunk 19]
```diff
@@ -2885,6 +2886,157 @@ def take(
...
+        Indices to change along each 1d slice of `arr`. This must match the
+        dimension of arr, but dimensions in Ni and Nj may be 1 to broadcast
+        against `arr`.
+    values : array_like (Ni..., J, Nk...)
+        values to insert at those indices. Its shape and dimension are
+        broadcast to match that of `indices`.
+    axis : int
+        The axis to take 1d slices along. If axis is None, the destination
+        array is treated as if a flattened 1d view had been created of it.
+        `axis=None` case is currenlty supported only for 1D input arrays.
```
Comment (manopapad) [inline cunumeric/module.py:3000]: typo: "currently"

File: cunumeric/module.py [hunk 20]
```diff
@@ -2885,6 +2886,157 @@ def take(
...
+
+    if a.ndim != indices.ndim:
+        raise ValueError(
+            "`indices` and `a` must have the same number of dimensions"
+        )
+    ind = _fill_fancy_index_for_along_axis_routines(
+        a.shape, computed_axis, indices
+    )
+    a[ind] = values
+    return
```
Comment (manopapad) [inline cunumeric/module.py:3037]: No need for a return statement here

File: cunumeric/module.py [hunk 21]
```diff
@@ -2545,7 +2545,8 @@ def take_along_axis(a: ndarray, indices: ndarray, axis: int) -> ndarray:
     Returns
     -------
     out: ndarray (Ni..., J, Nk...)
-        The indexed result.
+        The indexed result. It is going to be a vew to `arr` for most cases,
```
Comment (manopapad) [inline cunumeric/module.py:2937]: typo: "view"

**Review by manopapad — APPROVED** [review]
Some typos, otherwise looks good
